x/auth/keeper/msg_server_test.go

```diff
@@ -1,9 +1,88 @@
 package keeper_test
 
 import (
+	"github.com/golang/mock/gomock"
+
+	sdkmath "cosmossdk.io/math"
 	"cosmossdk.io/x/auth/types"
+	banktypes "cosmossdk.io/x/bank/types"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	simtestutil "github.com/cosmos/cosmos-sdk/testutil/sims"
+	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
+func (s *KeeperTestSuite) TestAsyncExec() {
+	addrs := simtestutil.CreateIncrementalAccounts(2)
+	coins := sdk.NewCoins(sdk.NewCoin("stake", sdkmath.NewInt(10)))
+
+	msg := &banktypes.MsgSend{
+		FromAddress: addrs[0].String(),
+		ToAddress:   addrs[1].String(),
+		Amount:      coins,
+	}
+
+	msgAny, err := codectypes.NewAnyWithValue(msg)
+	s.Require().NoError(err)
+
+	testCases := []struct {
+		name      string
+		req       *types.MsgNonAtomicExec
+		expectErr bool
+		expErrMsg string
+	}{
+		{
+			name: "empty signer address",
+			req: &types.MsgNonAtomicExec{
+				Signer: "",
+				Msgs:   []*codectypes.Any{},
+			},
+			expectErr: true,
+			expErrMsg: "empty signer address string is not allowed",
+		},
+		{
+			name: "invalid signer address",
+			req: &types.MsgNonAtomicExec{
+				Signer: "invalid",
+				Msgs:   []*codectypes.Any{},
+			},
+			expectErr: true,
+			expErrMsg: "invalid signer address",
+		},
+		{
+			name: "empty msgs",
+			req: &types.MsgNonAtomicExec{
+				Signer: addrs[0].String(),
+				Msgs:   []*codectypes.Any{},
+			},
+			expectErr: true,
+			expErrMsg: "messages cannot be empty",
+		},
+		{
+			name: "valid msg",
+			req: &types.MsgNonAtomicExec{
+				Signer: addrs[0].String(),
+				Msgs:   []*codectypes.Any{msgAny},
+			},
+			expectErr: false,
+		},
```

Comment:
we need to add the following test cases:
- multi msgs being executed
- multi msg with one failing being executed: the success state of the msg is kept, the failure state of the msg is discarded.
- multimsg and in one signer is different from the signer – send should fail, this is guaranteed by xaccounts send any, but if refactor happens we cna make sure this is not lost.